# Credit Risk Evaluator

In [1]:
# Import the dependencies: 
import numpy as np
import pandas as pd
from matplotlib import pyplot as plt
from pathlib import Path
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
from sklearn.linear_model import LogisticRegression
from sklearn.ensemble import RandomForestClassifier


## Retrieve the Data

The data is located in the Challenge Files Folder:

* `lending_data.csv`

Import the data using Pandas. Display the resulting dataframe to confirm the import was successful.

In [12]:
# Import the data
file_path = Path("Resources/lending_data.csv")
lending_df = pd.read_csv(file_path)
lending_df

loan_size  interest_rate  borrower_income  debt_to_income  \
0        10700.0          7.672            52800        0.431818   
1         8400.0          6.692            43600        0.311927   
2         9000.0          6.963            46100        0.349241   
3        10700.0          7.664            52700        0.430740   
4        10800.0          7.698            53000        0.433962   
...          ...            ...              ...             ...   
77531    19100.0         11.261            86600        0.653580   
77532    17700.0         10.662            80900        0.629172   
77533    17600.0         10.595            80300        0.626401   
77534    16300.0         10.068            75300        0.601594   
77535    15600.0          9.742            72300        0.585062   

       num_of_accounts  derogatory_marks  total_debt  loan_status  
0                    5                 1       22800            0  
1                    3                 0       13600            0  
2                    3                 0       16100            0  
3                    5                 1       22700            0  
4                    5                 1       23000            0  
...                ...               ...         ...          ...  
77531               12                 2       56600            1  
77532               11                 2       50900            1  
77533               11                 2       50300            1  
77534               10                 2       45300            1  
77535                9                 2       42300            1  

[77536 rows x 8 columns]

In [3]:
# Count of rows with null values
na_count = lending_df.isna().sum().sum()
print(f'There are {na_count} null values in the dataframe')
# 0 NA 
# Data types:
lending_df.dtypes
# Data are all float or int. 

There are 0 null values in the dataframe


loan_size           float64
interest_rate       float64
borrower_income       int64
debt_to_income      float64
num_of_accounts       int64
derogatory_marks      int64
total_debt            int64
loan_status           int64
dtype: object

## Predict Model Performance

- I think that the the logistic regression model will fit the data better than the Random Forests Classifier (RFC) model. 
  
- My guess is that since the outcome variable of loan_status (y) is binary (0 or 1) and there are only 7 features (X), the logit model will perform better than the RFC model. If there were a lot more features or if the outcome variable had more than two levels then I think that the RFC model would be better than the logit model.

## Split the Data into Training and Testing Sets

In [4]:
# Define the X (features) and y (target) sets:
X =  lending_df.drop("loan_status", axis=1)
y = lending_df["loan_status"]
# Check out the X (features) and y (target) sets:
# Will need to scale the data after splitting since the columns have very different value ranges:
print(X.head())
# y is good to runs since it is only 0 or 1:
y.unique()


   loan_size  interest_rate  borrower_income  debt_to_income  num_of_accounts  \
0    10700.0          7.672            52800        0.431818                5   
1     8400.0          6.692            43600        0.311927                3   
2     9000.0          6.963            46100        0.349241                3   
3    10700.0          7.664            52700        0.430740                5   
4    10800.0          7.698            53000        0.433962                5   

   derogatory_marks  total_debt  
0                 1       22800  
1                 0       13600  
2                 0       16100  
3                 1       22700  
4                 1       23000  


array([0, 1])

In [5]:
# Split the data into X_train, X_test, y_train, y_test
X_train, X_test, y_train, y_test = train_test_split(X, y, random_state=1)

### Scale the Features Data (X):

In [6]:
# Scale the independent variables, Create a StandardScaler() model and fit it to the training data
X_scaler = StandardScaler().fit(X_train)

In [7]:
# Transform the training and testing data by using the X_scaler
X_train_scaled = X_scaler.transform(X_train)
X_test_scaled = X_scaler.transform(X_test)

## Create, Fit and Compare Models

Create a Logistic Regression model, fit it to the data, and print the model's score. Do the same for a Random Forest Classifier. You may choose any starting hyperparameters you like. 

Which model performed better? How does that compare to your prediction? Write down your results and thoughts in the designated markdown cell.

### Train a Logistic Regression model and print the model score

In [8]:
# Train a Logistic Regression model and print the model score
# Create a logistic regresion model and train the logistic regression model
lr_model = LogisticRegression(max_iter=10000).fit(X_train_scaled, y_train)

### Logistic Regression model score:

In [9]:
# Print the training and testing scores
print(f"Training Data Score: {lr_model.score(X_train_scaled, y_train)}")
print(f"Testing Data Score: {lr_model.score(X_test_scaled, y_test)}")

Training Data Score: 0.9942908240473243
Testing Data Score: 0.9936545604622369


### Train a Random Forest Classifier model and print the model score

In [10]:
# Train a Random Forest Classifier model and fit the training data:
clf = RandomForestClassifier(random_state=1, n_estimators=50).fit(X_train_scaled, y_train)

### Random Forest Classifier model score

In [11]:
# Print the training and testing scores for the the Random Forests classifier model:
print(f'Training Score: {clf.score(X_train_scaled, y_train)}')
print(f'Testing Score: {clf.score(X_test_scaled, y_test)}')

Training Score: 0.9974893382858715
Testing Score: 0.9910751134956666


### Model Comparision and Write up:
- The logistic regression model performed just a bit better than the Random Forest Classifier model. The logisitic regression model had a test accuracy of 0.9937 and the Random Forest Classifier model had a test accuracy of 0.9911, so the logistic regression model was slightly more accurate. This was in line with my prediction that the logistic regression model will fit the data better than the RFC model, because of my prior reasoning that the outcome/target variable (loan_status)being binary would give the logistic model the better chance of performing with more accuracy.
  
- Both models were perhaps a bit overfit since their training scores were higher than their testing scores but the difference was not too far off.